design/src/rfcs/rfc0012_bytestream_callback_api.md

```diff
@@ -0,0 +1,112 @@
+RFC: A callback API for `ByteStream`
+====================================
+
+> Status: RFC
+
+Adding a callback API to `ByteStream` will enable developers using the SDK to implement things like checksum validations and 'read progress' callbacks.
+
+## The Implementation
+
+*Note that comments starting with '//' are not necessarily going to be included in the actual implementation and are intended as clarifying comments for the purposes of this RFC.*
+
+```rust
+// Each trait method defaults to doing nothing. It's up to implementors to
+// implement one or both of the trait methods
+/// Structs and enums implementing this trait can be inserted into a `ByteStream`,
+/// and will then be called in reaction to various events during a `ByteStream`'s
+/// lifecycle.
+pub trait ByteStreamReadCallback
+// I don't really like this bound but we need it because this will be inserted into an `Inner`
+where
+    Self: Debug + Clone + PartialEq + Eq,
```

Comment:
Would we want to include callbacks in the body equality checks? Seems like we should implement a custom equals comparison for ByteStream that excludes the callbacks.

If we `Arc` them in inner instead of `Box`, we can avoid the `Clone` bound.

Reply:
Yeah, I was being lazy mostly. I actually like `Arc` more because I think it's one way we could solve the issue of retrieving state from the callback struct.